File: 2022/day02a.py
#!/usr/bin/env python3

class Handshape:
    ROCK = 1
    PAPER = 2
    SCISSORS = 3
    def __init__(self, token):
        if token.lower() in ('a', 'x'):
            self.shape = self.ROCK
        elif token.lower() in ('b', 'y'):
            self.shape = self.PAPER
        elif token.lower() in ('c', 'z'):
            self.shape = self.SCISSORS
        else:
            raise ValueError("I don't know who you think you are or what you hope to accomplish here, but I haven't got time for it")
    def __repr__(self):
        if self.shape == self.ROCK:
            return "Rock"
        elif self.shape == self.PAPER:
            return "Paper"
        elif self.shape == self.SCISSORS:
            return "Scissors"
    def __gt__(self, other):
        return (self.shape == other.shape + 1) or (self.shape == self.ROCK and other.shape == other.SCISSORS)
    def __eq__(self, other):
        return self.shape == other.shape
    def value(self):
        return self.shape

def fight(opponent, me):
    points = me.value()
    if me == opponent:
        points += 3
    elif me > opponent:
        points += 6
    return points

if __name__ == "__main__":
    with open("day02.txt", "r") as f:
        data = f.read().splitlines()
    points = 0
    for line in data:
        opponent, me = map(lambda x: Handshape(x), line.split())
        points += fight(opponent, me)
    print(f"You've ended up with {points} points")

File: 2022/day02a.rs
use std::io::{BufReader, BufRead};
use std::fs::File;

fn fight(a:char, b:char) -> u32 {
    let other_hand = (a as u32) - 64;
    let my_hand = (b as u32) - 87;
    if my_hand == other_hand {
        return my_hand + 3;
    }
    else if (my_hand == other_hand + 1) || (my_hand == 1 && other_hand == 3) {
        return my_hand + 6;
    }
    else {
        return my_hand;
    }
}

fn main() {
    let reader = BufReader::new(File::open("day02.txt").unwrap());
    let mut my_score = 0;
    for line in reader.lines() {
        let line = line.unwrap();
        let mut iter = line.split_whitespace();
        let a = match iter.next() {
            Some(s) => s.chars().next().unwrap(),
            None => panic!("Malformed line?")
        };
        let b = match iter.next() {
            Some(s) => s.chars().next().unwrap(),
            None => panic!("Malformed line?")
        };
        my_score += fight(a, b);
    }
    println!("You've ended up with {} points", my_score);
}
